V9.3 External logins and 2FA for members (#3745)
umbraco/UmbracoDocs: Reference/Security/auto-linking/index.md

For some providers, it doesn't make sense to use auto-linking. This is especially true for public providers such as Google or Facebook. In those cases, it would mean that anyone who has a Google or Facebook account can log into your site. For public providers such as this, if auto-linking was needed you would need to limit the access by domain or other information provided in the Claims using the options/callbacks specified in those provider's authentication options.
:::

### Example for members
The way to implement auto linking for members is fairly similar to how it is for users. The main difference is the UI, where Umbraco do not have a fixed login page for members.
Instead, Umbraco ships with some Partial Macro Snippets for `Login` and `EditProfile` that contains handling of Login and manual linking of the configured external member providers.

When auto-linking is enabled, only the `Login` snippet is relevant as users do not have to register before.

The following example will show how to use Google and external login provider.
You can first create a `GoogleMemberExternalLoginProviderOptions.cs` file which configures the options like

```Csharp

> using Microsoft.Extensions.Options;

Reviewer: Can you add a `using System;` please because of the following `Array.Empty<string>` on line 41

                // Optionally specify the member groups names to add the auto-linking user to.
                defaultMemberGroups: Array.Empty<string>()
            )
            {
                // Optional callback
                OnAutoLinking = (autoLinkUser, loginInfo) =>
                {
                    // You can customize the member before it's linked.
                    // i.e. Modify the member's groups based on the Claims returned
                    // in the externalLogin info
                },
                OnExternalLogin = (user, loginInfo) =>
                {
                    // You can customize the member before it's saved whenever they have
                    // logged in with the external provider.
                    // i.e. Sync the member's name based on the Claims returned
                    // in the externalLogin info

                    return true; //returns a boolean indicating if sign in should continue or not.
                }
            };
    }
}
```

To register this configuration class, you can call the following from your `startup.cs`:
```Csharp
services.ConfigureOptions<GoogleMemberExternalLoginProviderOptions>();
```

> Like for users, we recommend to create an extension method on the `IUmbracoBuilder`, to add the Google Authentication, like this
> This extension can also handle the configuration of `GoogleMemberExternalLoginProviderOptions`:

Reviewer 2: ```suggestion
Like for users, we recommend creating an extension method on the `IUmbracoBuilder`, to add the Google Authentication, like this.
This extension can also handle the configuration of `GoogleMemberExternalLoginProviderOptions`:
```

```Csharp
public static IUmbracoBuilder AddMemberGoogleAuthentication(this IUmbracoBuilder builder)
{
    // Register GoogleMemberExternalLoginProviderOptions here rather than require it in startup
    builder.Services.ConfigureOptions<GoogleMemberExternalLoginProviderOptions>();

    builder.AddMemberExternalLogins(logins =>
    {
        logins.AddMemberLogin(
            memberAuthenticationBuilder =>
            {
                memberAuthenticationBuilder.AddGoogle(
                    // The scheme must be set with this method to work for the umbraco members
                    memberAuthenticationBuilder.SchemeForMembers(GoogleMemberExternalLoginProviderOptions.SchemeName),
                    options =>
                    {
                        options.ClientId = "YOURCLIENTID";
                        options.ClientSecret = "YOURCLIENTSECRET";
                    });
            });
    });
    return builder;
}
```

Finally this extension can also be called from the `Startup.cs` like the example below:

```Csharp
services.AddUmbraco(_env, _config)
   .AddBackOffice()
   .AddWebsite()
   .AddComposers()
   .AddMemberGoogleAuthentication()
   .Build();
```
:::note

> Auto-linking only make sense if you have a public member registration anyway or the external provider do not have public account creation.

Reviewer 2: ```suggestion
Auto-linking only makes sense if you have a public member registration anyway or the external provider does not have public account creation.
```

:::

## Local logins

> If you have configured auto-linking, then any auto-linked user/member will have an empty password assigned and they will not be able to log in locally (via username and password). In order to log in locally, they will have to assign a password to their account in the backoffice or the edit profile page.

Reviewer 2: ```suggestion
If you have configured auto-linking, then any auto-linked user or member will have an empty password assigned and they will not be able to log in locally (via username and password). In order to log in locally, they will have to assign a password to their account in the backoffice or the edit profile page.
```
